### ISSUE TITLE
Anaconda cloud package

### ISSUE THREAD

**Issue by jhrmnn** [i-pi/i-pi#16]
Would there be an interest to have i-PI available on the Anaconda cloud? I'd be happy to setup a CI on Travis that would build the packages for macOS and Linux and deploy to the cloud.

**Comment by jhrmnn** [comment]
Nevermind. I haven't realized at first that the compilation with Make is only necessary for the example drivers, while i-PI itself is a pure Python package. Any reason why installation with Pip is not recommended in the Readme?

```
pip install git+https://github.com/i-pi/i-pi.git
```

And any reason why the i-PI socket library is compiled by `setup.py` when it's actually not used by the i-PI package?

**Comment by ceriottm** [comment]
I think it'd be a great idea to recommend a pip installation, and I think the reason the socket library is compiled by setup.py is probably because whoever wrote setup.py (possibly me) wanted to try and see how one would include a compile task in it. Feel free to update setup.py so that it's more functional and has a bumped up version number!

**Comment by jhrmnn** [comment]
I'll be also happy to put i-PI on PyPI, and transfer ownership to some core developer.

**Comment by mahrossi** [comment]
I agree with this idea.

**Comment by jhrmnn** [comment]
I'll create the package from the current `v2` branch, so it will be one commit ahead of the `v2.0.0` tag, I guess that's ok.

What should go in the package?

- What's the relation of `ipi/tests/` and `ipi_tests/`? I guess the later is useful only for development and shouldn't be in the package, but what about the former? What about `tests/`—looks just like some leftover.
- I guess `licences/` should go in.
- What about all the symbolic links in `bin/` that point to `../tools/py/*`, are they actively being used? If so, I'll include them in `setup.py` so they get installed properly.
- What about stuff in `tools/` that isn't symlinked from `bin/`—is that useful to users or only to developers?

**Comment by ceriottm** [comment]
Eh. TBH, all of these things you mention are us doing a poor job when preparing the release. Basically each of those items would require a discussion, and some boring and unrewarding clean-up work.... 
It'd be great to have @OndrejMarsalek chime into the discussion, as perhaps we should make a small clean-up commit and cherry-pick it into v2.0 before we go with creating the pipy package

**Comment by sirmarcel** [comment]
Hello everyone! Any news on this effort?

**Comment by jhrmnn** [comment]
My answers above were not answered, so I haven't followed up on this.

**Comment by ceriottm** [comment]
Let's try and revive this thread - I'd really like to get @OndrejMarsalek to react as he also had quite clear opinions on several clean-ups that needed to be done.

**Comment by mahrossi** [comment]
Just looked at this thread. We should definitely revive it for v3 if @jhrmnn still has interest

**Comment by jhrmnn** [comment]
Sure, I'll be happy to put i-PI on PyPI. Has anything changed regarding the code structure since my comment from February?

**Comment by mahrossi** [comment]
Re. your questions:

> I'll create the package from the current `v2` branch, so it will be one commit ahead of the `v2.0.0` tag, I guess that's ok.
> 
> What should go in the package?
> 
> * What's the relation of `ipi/tests/` and `ipi_tests/`? I guess the later is useful only for development and shouldn't be in the package, but what about the former? What about `tests/`—looks just like some leftover.

I think there is no harm in keeping `ipi/tests` in but TBH I had never myself looked into `ipi_tests/`. Maybe Michele can comment. Looks to me like some old stuff. `tests/` definitely can go away and the script inside moved to somewhere else in the future, but anyways, it is not important.

> * I guess `licences/` should go in.

Yes

> * What about all the symbolic links in `bin/` that point to `../tools/py/*`, are they actively being used? If so, I'll include them in `setup.py` so they get installed properly.

Some are more for dev purposes (e.g. `i-pi-pepper`) and others actively used but all can stay in, really. 

> * What about stuff in `tools/` that isn't symlinked from `bin/`—is that useful to users or only to developers?

That is useful for users for sure. All of that should go in. Some pretty necessary post-processing scripts are there that are not symlinked (could be made though, will check that in the future).

> * I won't include `doc/`, `drivers/`, and `examples/` in the package.

Drivers could be useful to include. Other opinions welcome.

**Comment by ceriottm** [comment]
On Tue, 5 Nov 2019 at 08:25, Mariana Rossi <notifications@github.com> wrote:

> Re. your questions:
>
> I'll create the package from the current v2 branch, so it will be one
> commit ahead of the v2.0.0 tag, I guess that's ok.
>
> What should go in the package?
>
>    - What's the relation of ipi/tests/ and ipi_tests/? I guess the later
>    is useful only for development and shouldn't be in the package, but what
>    about the former? What about tests/—looks just like some leftover.
>
> I think there is no harm in keeping ipi/tests in but TBH I had never
> myself looked into ipi_tests/. Maybe Michele can comment. Looks to me
> like some old stuff. tests/ definitely can go away and the script inside
> moved to somewhere else in the future, but anyways, it is not important.
>

We need to streamline tests and regtests, but I don't know whether and when
that will happen.


>
>    - I guess licences/ should go in.
>
> Yes
>
>
>    - What about all the symbolic links in bin/ that point to ../tools/py/*,
>    are they actively being used? If so, I'll include them in setup.py so
>    they get installed properly.
>
> Some are more for dev purposes (e.g. i-pi-pepper) and others actively
> used but all can stay in, really.
>
>
>    - What about stuff in tools/ that isn't symlinked from bin/—is that
>    useful to users or only to developers?
>
> That is useful for users for sure. All of that should go in. Some pretty
> necessary post-processing scripts are there that are not symlinked (could
> be made though, will check that in the future).
>
I think that all tools should be symlinked. And actually, we should work on
making the options and interface more consistent across tools. Another TODO
item for when the hell freezes solid....

>
>    - I won't include doc/, drivers/, and examples/ in the package.
>
> I would include everything, the examples are very useful. Is it possible
to have a separate i-pi-doc package that includes doc and examples?
I guess that problem including drivers is that they require compilation.
How is this generally done?

**Comment by jhrmnn** [comment]
> > * What about stuff in `tools/` that isn't symlinked from `bin/`—is that useful to users or only to developers?
> 
> That is useful for users for sure. All of that should go in. Some pretty necessary post-processing scripts are there that are not symlinked (could be made though, will check that in the future).

Ok, should I then include *all* the scripts from `/tools/py` while removing the `.py` extension and replacing underscores with hyphens (just for the package installation, not in the repo)?

`/tools/bash/` and `/tools/f90/` can be omitted I guess?

> > * I won't include `doc/`, `drivers/`, and `examples/` in the package.
>
> Drivers could be useful to include. Other opinions welcome.

> I guess that problem including drivers is that they require compilation. How is this generally done?

Not only that they require compilation, but they are actually a standalone program. The Python package ecosystem is really not supposed to handle that. Compiled Python extensions are always just dynamic libraries that are actually used from within the Python application.

Conda packages should be able to handle this, but I have no experience with this particular use case.

In general, I would suggest to release the driver separately as a a source distribution ( `.tar.gz` file) on Github, which can be compiled by downloading, untaring, and running `make` (or `./configure && make && make install` if you want to be fancy). This is the simplest and straightforward no-questions-asked solution.

> We need to streamline tests and regtests, but I don't know whether and when that will happen.

> And actually, we should work on making the options and interface more consistent across tools.

Do you mean you want to wait with the Pip package before this gets done?

> to have a separate i-pi-doc package that includes doc and examples?

The files of an installed Python package are not easily accessible to the user, so this wouldn't be too useful. Would it be too much work to turn the docs into a website and serve it via Github Pages?

**Comment by mahrossi** [comment]
> > > * What about stuff in `tools/` that isn't symlinked from `bin/`—is that useful to users or only to developers?
> > 
> > 
> > That is useful for users for sure. All of that should go in. Some pretty necessary post-processing scripts are there that are not symlinked (could be made though, will check that in the future).
> 
> Ok, should I then include _all_ the scripts from `/tools/py` while removing the `.py` extension and replacing underscores with hyphens (just for the package installation, not in the repo)?
> 
> `/tools/bash/` and `/tools/f90/` can be omitted I guess?
> 
> > > * I won't include `doc/`, `drivers/`, and `examples/` in the package.
> > 
> > 
> > Drivers could be useful to include. Other opinions welcome.
> 
> > I guess that problem including drivers is that they require compilation. How is this generally done?
> 
> Not only that they require compilation, but they are actually a standalone program. The Python package ecosystem is really not supposed to handle that. Compiled Python extensions are always just dynamic libraries that are actually used from within the Python application.
> 
> Conda packages should be able to handle this, but I have no experience with this particular use case.
> 
> In general, I would suggest to release the driver separately as a a source distribution ( `.tar.gz` file) on Github, which can be compiled by downloading, untaring, and running `make` (or `./configure && make && make install` if you want to be fancy). This is the simplest and straightforward no-questions-asked solution.
> 

I like this solution and on our side it is very easy to implement. If there are no voices against it, I can create a `driver` repo in the `i-pi` github "now" including such a tar file.

> > We need to streamline tests and regtests, but I don't know whether and when that will happen.
> 
> > And actually, we should work on making the options and interface more consistent across tools.
> 
> Do you mean you want to wait with the Pip package before this gets done?

I'll let MC answer that.


> 
> > to have a separate i-pi-doc package that includes doc and examples?
> 
> The files of an installed Python package are not easily accessible to the user, so this wouldn't be too useful. Would it be too much work to turn the docs into a website and serve it via Github Pages?

After reading I realized that I agree with Michele that `examples` are very useful. Can those be included without pain? Regarding the documentation, it would be wonderful to have it as a website and I have no idea at the moment how exactly to do it or who to ask for it... :-D

**Comment by jhrmnn** [comment]
> I like this solution and on our side it is very easy to implement. If there are no voices against it, I can create a driver repo in the i-pi github "now" including such a tar file.

Don't know about your workflow here, but usually I find it easier to keep everything in a single repo. Especially if they are logically related. Say you change API in i-PI, you'll need to also change the driver. Then it makes sense if all these things are in a single commit. So I'd actually suggest to keep everything in the `/drivers/` directory as it is now, and basically just `tar` that.

> After reading I realized that I agree with Michele that examples are very useful. Can those be included without pain? Regarding the documentation, it would be wonderful to have it as a website and I have no idea at the moment how exactly to do it or who to ask for it... :-D

As I wrote above, the problem is that when a user installs a Python package, they don't have any normal way to access the files. That's why it doesn't make much sense to include the examples into the PyPI package IMO.

I see that your docs are Latex. So I would suggest to, besides the `drivers.tar.gz` file, also release (on Github releases) a `docs.tar.gz` file which will contain all the examples and the compiled PDF manual.

So in the end, if a user wants to get the full thing, they will run `pip install ipi` and download and untar two files from Github releases. This sounds perfectly fine to me.

**Comment by jhrmnn** [comment]
Or even release the compiled manual and the examples as two separate files, so once can have a look at the manual with a single click.

**Comment by jhrmnn** [comment]
> Regarding the documentation, it would be wonderful to have it as a website and I have no idea at the moment how exactly to do it or who to ask for it... :-D

Looking at how i-PI documentation is generated, it seems to me that converting it to a website form would be a major effort.

**Comment by ceriottm** [comment]
Yeah, I was young and unaware of modern code documentation technology....

I wouldn't wait for everything to be perfect to release a pip version.
Actually I'd move quickly so we have a py2 version,
before we go ahead with the py3 conversion that is planned to happen in the
next few weeks

**Comment by jhrmnn** [comment]
Turns out that the scripts cannot be renamed during the package installation. Can I safely rename them directly in the repository? At which point one could also think about actually moving them to the `/bin/` directory.

**Comment by ceriottm** [comment]
In the repo you can have symlinks with a different name. I like the idea of
the scripts to reside in a tools/XX/ folder, and to have i-pi-YYY names à
la Gromacs only when they're called. You can't have symlinks in the package?

**Comment by jhrmnn** [comment]
Right. Unfortunately not, because the package installer puts the scripts and the package files in different locations. Usually, this would be something like `lib/python2.7/site-packages/ipi` for the actual package and `bin/` for the scripts (so that they are in the user's `PATH`), but it could be something else, depending on how was the user's Python distribution (or virtual environment) set up.

**Comment by jhrmnn** [comment]
I just stumbled upon this again. I went ahead and made a package from current master and put it on test PyPI:

https://test.pypi.org/project/i-PI/

I've included the only the `ipi` directory, `licenses`, and `bin/i-pi`.

Feedback?

**Comment by jhrmnn** [comment]
Let's talk dependencies.

- Do we know the lowest version of Numpy that works?
- I see use of Scipy in 4 files in `ipi/engine/motion` and in 2 files in `ipi/utils`, is Scipy required?
- Besides Scipy, I see:
    - `pyfftw` in `ipi/utils/nmtransform.py`
    -  `plumed`, `yaff`, and `sgdml` in `ipi/engine/forcefields.py`
    -  `pyscenarios` in `ipi/engine/motion/scphonons.py`

    Should some of them be perhaps installed as extras (`pip install i-pi[<some extras label>,<some other extras label>]`)

**Comment by mahrossi** [comment]
Hi Jan,
All packages including scipy are, in principle, optional extras - meaning that the i-PI will run if you have none of them but some features may not be available.

For the recommended versions of the packages you ask, I think the current workflow of the CI is what we guarantee but not sure we know the oldest version of numpy. 

And finally, not sure it is a pain or not, but I would also include at least the "drivers" folder in the package.

**Comment by jhrmnn** [comment]
The extras are essentially just an easy way for users to install optional dependencies. Usually, the extras labels correspond to some functionality. But if in i-PI's case, each individual optional dependency presents a different functionality, then I'd just not define any extras and the user would just install the optional dependency directly. What do you think?

The CI runs at the latest version, 1.20. I can do that, but it seems too restrictive—people might use Numpy in environments with older Numpy and this would force them to update Numpy. Numpy is relatively stable, and unless you're using some cutting-edge Numpy functionality, I'd guess i-PI works fine with older versions. So I'd suggest to go with 1.12 from Jan 2017, and we can always increase it if someone reports an issue.

As for the `drivers`—sure, I can include it, but the question is what is it supposed to do in the package? When you install the a package with Pip, the files end up somewhere, most people don't know where. If the user actually wants to use the drivers, then need to have access to the files.

**Comment by jhrmnn** [comment]
Maybe even more open-minded—let's just specify Numpy without a version pin, and restrict in the future if someone reports issues.

**Comment by jhrmnn** [comment]
And one final decision—should I create a version based on the last release, 2.3, or a prerelease version based on current `master`, with first stable release once 2.4 (or 3.0?) is out?

**Comment by jhrmnn** [comment]
@mahrossi pointed out there is a pure-Python driver, so it makes sense to include that in the package, will do.
